dttools/src/mq.c

```diff
@@ -0,0 +1,716 @@
+#include <unistd.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <fcntl.h>
+#include <signal.h>
+#include <poll.h>
+#include <sys/socket.h>
+#include <arpa/inet.h>
+
+#include "mq.h"
+#include "buffer.h"
+#include "list.h"
+#include "itable.h"
+#include "set.h"
+#include "link.h"
+#include "xxmalloc.h"
+#include "debug.h"
+#include "jx_print.h"
+#include "jx_parse.h"
+
+
+#define HDR_SIZE (sizeof(struct mq_msg) - offsetof(struct mq_msg, magic))
+#define HDR_MAGIC "MQmsg"
+
+enum mq_msg_type {
+	MQ_MSG_BUFFER = 1,
+	MQ_MSG_JSON = 2,
+};
+
+enum mq_socket {
+	MQ_SOCKET_SERVER,
+	MQ_SOCKET_INPROGRESS,
+	MQ_SOCKET_CONNECTED,
+	MQ_SOCKET_ERROR,
+};
+
+struct mq_msg {
+	size_t len;
+	bool parsed_header;
+	ptrdiff_t hdr_pos;
+	ptrdiff_t buf_pos;
+	buffer_t buf;
+	struct jx *j;
+
+	/* Here be dragons!
+	 *
+	 * Since we need to be able allow send/recv to be interrupted at any time
+	 * (even in the middle of an int), we can't rely on reading/writing
+	 * multi-byte header fields all in one go. The following fields are
+	 * arranged to match the wire format of the header. DO NOT add additional
+	 * struct fields below here! (If you do change the header format, be sure
+	 * to update the sanity check in msg_create.) First is some padding
+	 * (void pointer is self-aligned) so we don't need to worry too much
+	 * about the alignment of the earlier fields. Then the actual header
+	 * follows. Note that the length in the header needs to be in network
+	 * byte order, so that gets stored separately.
+	 *
+	 *  0 1 2 3 4 5 6 7
+	 * +-+-+-+-+-+-+-+-+
+	 * |  magic  |pad|*|    *type
+	 * +-+-+-+-+-+-+-+-+
+	 * |    length     |
+	 * +-+-+-+-+-+-+-+-+
+	 */
+	void *pad1;
+
+	char magic[5];
+	char pad2[2]; // necessary for alignment, should be 0
+	uint8_t type;
+	uint64_t hdr_len;
+};
+
+struct mq {
+	struct link *link;
+	enum mq_socket state;
+	struct mq *acc;
+	struct list *send;
+	int err;
+	struct mq_msg *recv;
+	struct mq_msg *send_buf;
+	struct mq_msg *recv_buf;
+	struct mq_poll *poll_group;
+};
+
+struct mq_poll {
+	struct itable *members;
+	struct set *acceptable;
+	struct set *readable;
+	struct set *error;
+};
+
+
+#ifndef htonll
+static uint64_t htonll(uint64_t hostlonglong) {
+	uint64_t out = 0;
+	uint8_t *d = (uint8_t *) &out;
+	d[7] = hostlonglong>>0;
+	d[6] = hostlonglong>>8;
+	d[5] = hostlonglong>>16;
+	d[4] = hostlonglong>>24;
+	d[3] = hostlonglong>>32;
+	d[2] = hostlonglong>>40;
+	d[1] = hostlonglong>>48;
+	d[0] = hostlonglong>>56;
+	return out;
+}
+#endif
+
+#ifndef ntohll
+static uint64_t ntohll(uint64_t netlonglong) {
+	uint64_t out = 0;
+	uint8_t *d = (uint8_t *) &netlonglong;
+	out |= (uint64_t) d[7]<<0;
+	out |= (uint64_t) d[6]<<8;
+	out |= (uint64_t) d[5]<<16;
+	out |= (uint64_t) d[4]<<24;
+	out |= (uint64_t) d[3]<<32;
+	out |= (uint64_t) d[2]<<40;
+	out |= (uint64_t) d[1]<<48;
+	out |= (uint64_t) d[0]<<56;
+	return out;
+}
+#endif
+
+static int ppoll_compat(struct pollfd fds[], nfds_t nfds, int stoptime) {
```

Comment:
Does this compat implementation have a race condition, compared to native `ppoll`?
If so, we can't avoid it, but it should probably be documented.
Perhaps move to `dttools/src/ppoll_compat.c`?

Reply:
It does, I can move that to its own module